shell/renderer/electron_renderer_client.cc

```diff
@@ -41,117 +41,122 @@ void ElectronRendererClient::RenderFrameCreated(
 void ElectronRendererClient::RunScriptsAtDocumentStart(
     content::RenderFrame* render_frame) {
   RendererClientBase::RunScriptsAtDocumentStart(render_frame);
+
   // Inform the document start phase.
-  v8::HandleScope handle_scope(v8::Isolate::GetCurrent());
+  auto* isolate = v8::Isolate::GetCurrent();
+  v8::HandleScope handle_scope(isolate);
   node::Environment* env = GetEnvironment(render_frame);
   if (env)
-    gin_helper::EmitEvent(env->isolate(), env->process_object(),
-                          "document-start");
+    gin_helper::EmitEvent(isolate, env->process_object(), "document-start");
 }
 
 void ElectronRendererClient::RunScriptsAtDocumentEnd(
     content::RenderFrame* render_frame) {
   RendererClientBase::RunScriptsAtDocumentEnd(render_frame);
+
   // Inform the document end phase.
-  v8::HandleScope handle_scope(v8::Isolate::GetCurrent());
+  auto* isolate = v8::Isolate::GetCurrent();
+  v8::HandleScope handle_scope(isolate);
   node::Environment* env = GetEnvironment(render_frame);
   if (env)
-    gin_helper::EmitEvent(env->isolate(), env->process_object(),
-                          "document-end");
+    gin_helper::EmitEvent(isolate, env->process_object(), "document-end");
 }
 
 void ElectronRendererClient::DidCreateScriptContext(
-    v8::Handle<v8::Context> renderer_context,
+    v8::Handle<v8::Context> context,
     content::RenderFrame* render_frame) {
   // TODO(zcbenz): Do not create Node environment if node integration is not
   // enabled.
 
   // Only load Node.js if we are a main frame or a devtools extension
   // unless Node.js support has been explicitly enabled for subframes.
-  if (!ShouldLoadPreload(renderer_context, render_frame))
+  if (!ShouldLoadPreload(context, render_frame))
     return;
 
   injected_frames_.insert(render_frame);
 
   if (!node_integration_initialized_) {
     node_integration_initialized_ = true;
-    node_bindings_->Initialize(renderer_context);
+    node_bindings_->Initialize(context);
     node_bindings_->PrepareEmbedThread();
   }
 
   // Setup node tracing controller.
   if (!node::tracing::TraceEventHelper::GetAgent())
     node::tracing::TraceEventHelper::SetAgent(node::CreateAgent());
 
-  // Setup node environment for each window.
-  v8::Maybe<bool> initialized = node::InitializeContext(renderer_context);
+  v8::Maybe<bool> initialized = node::InitializeContext(context);
   CHECK(!initialized.IsNothing() && initialized.FromJust());
 
-  node::Environment* env =
-      node_bindings_->CreateEnvironment(renderer_context, nullptr);
+  // If DidCreateScriptContext is called and we've already created a Node.js
+  // Environment, then we're in the same process with a new V8::Context. If
+  // render_frame is the main frame, then we should replace it. Otherwise, we
+  // should assign the existing Environment to the new V8::Context.
+  if (env_ && !render_frame->IsMainFrame()) {
```

Comment:
The frame in child window opened by `window.open` is also a main frame, with this change there are still multiple environments created after calling `window.open`, it is just the new code no longer does cleanup for the old environment.

Another case is when you reload a page, this code still creates a new environment for the new context, and does not clean up the old one.

Reply:
> Can you point where is this handled in chromium ?

I don't have a reference to the code, this reloading behavior caused some very nasty bugs in `remote` module so it burnt into my memory 😕. You can find some tracings of how js contexts get released during reloads in this issue: https://github.com/electron/electron/issues/8205#issuecomment-403377520.